apps/inheritance_app/inheritance_encrypt_data.c

```diff
@@ -468,23 +597,91 @@ static bool encrypt_data(void) {
+static bool get_pb_encoded_buffer(
+    const inheritance_encrypt_data_with_pin_encrypted_data_structure_t *result,
+    uint8_t *buffer,
+    uint16_t max_buffer_len,
+    size_t *bytes_written_out) {
+  if (NULL == result || NULL == buffer || NULL == bytes_written_out) {
+    return false;
+  }
+  /* Create a stream that will write to our buffer. */
+  pb_ostream_t stream = pb_ostream_from_buffer(buffer, max_buffer_len);
+
+  /* Now we are ready to encode the message! */
+  bool status = pb_encode(
+      &stream,
+      INHERITANCE_ENCRYPT_DATA_WITH_PIN_ENCRYPTED_DATA_STRUCTURE_FIELDS,
+      result);
+
+  if (true == status) {
+    *bytes_written_out = stream.bytes_written;
+  }
+
+  return status;
+}
+
+static bool inheritance_send_in_chunks(inheritance_query_t *query,
+                                       const uint8_t *buffer,
+                                       const size_t buffer_len) {
+  size_t total_count = ((buffer_len % ENCRYPTED_CHUNK_SIZE) > 0)
+                           ? (buffer_len / ENCRYPTED_CHUNK_SIZE) + 1
+                           : (buffer_len / ENCRYPTED_CHUNK_SIZE);
+  size_t remaining_size = (size_t)buffer_len;
+  size_t offset = 0;
   inheritance_result_t result =
       init_inheritance_result(INHERITANCE_RESULT_ENCRYPT_TAG);
   result.encrypt.which_response =
-      INHERITANCE_ENCRYPT_DATA_WITH_PIN_RESPONSE_RESULT_TAG;
-  if (!inheritance_get_query(query, INHERITANCE_QUERY_ENCRYPT_TAG) ||
-      !check_which_request(
-          query, INHERITANCE_ENCRYPT_DATA_WITH_PIN_REQUEST_INITIATE_TAG)) {
-    return false;
+      INHERITANCE_ENCRYPT_DATA_WITH_PIN_RESPONSE_ENCRYPTED_DATA_TAG;
+  result.encrypt.encrypted_data.chunk_payload.chunk_index = 0;
+  result.encrypt.encrypted_data.chunk_payload.total_chunks = total_count;
+
+  for (int index = 0; index < total_count; index++) {
+    if (!inheritance_get_query(query, INHERITANCE_QUERY_ENCRYPT_TAG) ||
+        !check_which_request(
+            query,
+            INHERITANCE_ENCRYPT_DATA_WITH_PIN_REQUEST_ENCRYPTED_DATA_REQUEST_TAG)) {
+      return false;
+    }
+    // chunk_payload validation checks
+    if (query->encrypt.encrypted_data_request.has_chunk_ack == false ||
+        query->encrypt.encrypted_data_request.chunk_ack.chunk_index != index) {
+      return false;
+    }
+    size_t chunk_size = (remaining_size > ENCRYPTED_CHUNK_SIZE)
+                            ? ENCRYPTED_CHUNK_SIZE
+                            : remaining_size;
+    remaining_size -= chunk_size;
+    result.encrypt.encrypted_data.chunk_payload.remaining_size = remaining_size;
+    result.encrypt.encrypted_data.has_chunk_payload = true;
+    memcpy(result.encrypt.encrypted_data.chunk_payload.chunk.bytes,
+           buffer + offset,
+           chunk_size);
+    result.encrypt.encrypted_data.chunk_payload.chunk.size = chunk_size;
+    inheritance_send_result(&result);
+    offset += chunk_size;
+    result.encrypt.encrypted_data.chunk_payload.chunk_index++;
```

Comment:
Instead of separately incrementing `chunk_index` you can just use this as the iterator in for loop, that way we have one less variable to worry about